File: leetcode/array/2.addTwoNumbers/2addTwoNumbers.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummyHead = ListNode(0)
        p = l1
        q = l2
        currentNode = dummyHead
        carry = 0  # 向前的进位
        while p or q:
            x = p.val if p else 0  # 取值，若当前节点唯空，取值为0
            y = q.val if q else 0
            sum = carry + x + y
            carry = sum // 10  # compute new carry
            currentNode.next = ListNode(sum % 10)  # get new node
            currentNode = currentNode.next
            p = p.next if p else None  
            q = q.next if q else None
            
        # end_while
        if carry > 0:
            currentNode.next = ListNode(carry)
        
        return dummyHead.next

File: leetcode/array/2.addTwoNumbers/2addTwoNumbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */

var addTwoNumbers = function(l1, l2) {
    let carry = 0;
    let result = new ListNode(0);
    let currentResult,
        currentNode = result;
    while (l1 || l2 || carry) {
        if (!l2) {
            l2 = new ListNode(0);
        }
        if (!l1) {
            l1 = new ListNode(0);
        }
        currentResult = carry + l1.val + l2.val;
        if (currentResult / 10 >= 1) {
            carry = 1;
            currentNode.next = new ListNode(currentResult % 10);
        } else {
            currentNode.next = new ListNode(currentResult);
            carry = 0;
        }
        currentNode = currentNode.next;
        l1 = l1.next;
        l2 = l2.next;
    }

    return result.next;
};
